### PR TITLE
Bump Qt requirement to 5.12

### PR DESCRIPTION
**Pull request by Scrumplex**
This PR drops support for Qt versions below 5.12.

Also: This bumps minimum CMake version to 3.15, as QuaZip requires this as well.

This effectively means that the following platforms are supported:
- Alpine 3.11 or later
- Arch
- Debian 11 (bullseye) or later
- Fedora 30 or later
- macOS 10.12 (Sierra) or later
- NixOS 21.05 or later
- openSUSE Leap 15.2
- openSUSE Tumbleweed
- Ubuntu 20.04 or later
- Void
- Windows 7 or later

Notable unsupported platforms:
- Debian 10 (doesn't support Qt 5.12)
- Ubuntu 18.04 (doesn't support Qt 5.12 nor CMake 3.15)

NOTE: This does not mean that it will be impossible to run PolyMC on those platforms. A packager could patch PolyMC to build on a slightly older Qt or just use the AppImage / Flatpak

### CODE REVIEW

**Comment by ghost** [comment]
I don't really like having Qt 5.12 required.

The main point of PolyMC is to expand support to as many users and platforms as possible. Qt 5.12 cuts off several lines of OSes that are otherwise pretty widely used. As I was saying Qt 5.9 is the newest that should really be required. Anything newer would cut off a chunk of the userbase.

**Comment by Scrumplex** [comment]
> The main point of PolyMC is to expand support to as many users and platforms as possible. Qt 5.12 cuts off several lines of OSes that are otherwise pretty widely used

Actually Debian 10 is the only distro that's cut off because of Qt 5.12. Ubuntu 18.04 doesn't even support CMake 3.15, which is required for QuaZip. In both cases you can either use a PPA for Qt (and CMake) or use the other distribution options (AppImage or Flatpak). Maybe we could also provide a portable binary for Linux that brings it's own Qt 5.12

**Comment by Scrumplex** [comment]
arrgh! Windows

**Comment by Scrumplex** [comment]
btw. Let's not merge this for 1.1.0

**Comment by ghost** [comment]
Right, I was talking about MacOS and Windows.

I personally still have a Mac that only goes up to Mavericks, and I imagine many others do as well. Unless we can maintain a separate branch with support for older versions, Qt 5.12 shouldn't be required *yet*

**Comment by glowiak** [comment]
Please don't do that. I was to lazy to build newer qt on my Slack 14.2 and I'm stuck with AlienBOB's qt5.9.6

**Comment by Scrumplex** [comment]
> Please don't do that. I was to lazy to build newer qt on my Slack 14.2 and I'm stuck with AlienBOB's qt5.9.6

As far as I can tell you are already unsupported regardless of this PR, as CMake 3.15 is required for the QuaZip dependency. It seems like you would need to use a newer Slackware (15?) anyway to build this in it's currrent state: https://repology.org/project/cmake/versions

**Comment by Scrumplex** [comment]
@glowiak In your PR #131 you linked to http://slackbuilds.org/repository/14.2/libraries/qt5/

It says Qt 5.12 so isn't this covered then?

**Comment by Scrumplex** [comment]
> Right, I was talking about MacOS and Windows.
> 
> I personally still have a Mac that only goes up to Mavericks, and I imagine many others do as well. Unless we can maintain a separate branch with support for older versions, Qt 5.12 shouldn't be required _yet_

Apart from Mavericks being 8-9 years old, I don't know if we should target unsupported OSes at all. I mean should we target Windows Vista or XP? I know they are even older, but Windows just has a slower release cycle.

Apart from that I feel like we have a few dependencies that already block older OSes. Notably QuaZip requires CMake 3.15 to build. Oracle JDK 17 only supports macOS 10.14 and newer.

Of course some people might wanna play older versions of Minecraft, but then they could just use an older version of this launcher. Support will end eventually. I personally prefer sooner than later when talking about EOL platforms.

About old Mavericks Macs: What about putting Linux on them? They are probably well supported by now.

**Comment by d-513** [comment]
If qt5.12 is not viable, please bump at least to 5.7. 5.7 introduced QApplication:setDesktopFileName, which should fix the wayland icon issues, while having the desktop file capitalized normally

**Comment by Scrumplex** [comment]
> If qt5.12 is not viable, please bump at least to 5.7. 5.7 introduced QApplication:setDesktopFileName, which should fix the wayland icon issues, while having the desktop file capitalized normally

Why is it not viable? The only counterargument is that we drop support for platforms, that are not even supported by their vendors. Why do we care about old macOS versions that haven't received security updates for years? When a platform reaches end of life, there is no reason to support it anymore. 

- Legacy macOS users can (and should) switch to modern Linux
  - Legacy macOS is insecure, the same way Windows XP or Vista is.
  - Modern macOS applications don't run on older releases anyway, so you don't get anything by sticking to macOS as a user
- Ubuntu 18.04 users can (and should) upgrade to 20.04.
- Debian 10 users can (and should) upgrade to 11.

It makes the codebase inherently more complex to support modern and legacy platforms at the same time. I don't see value in supporting old platforms, as there is an upgrade path for all of them.

**Comment by d-513** [comment]
@Scrumplex I agree with you, there have been opinions that this shouldnt be merged so my "not viable" was more like "if this is rejected".

**Comment by Scrumplex** [comment]
> @Scrumplex I agree with you, there have been opinions that this shouldnt be merged so my "not viable" was more like "if this is rejected".

Oh, sorry about that. I didn't notice the `if` in `If qt5.12 is not viable` :sweat_smile:

**Comment by DioEgizio** [comment]
debian 10 won't work for a qt issue anyways (see https://bugreports.qt.io/browse/QTBUG-68393)

so we would only miss ubuntu 18.04 (why would you use it anyways), windows xp and vista (just why) and older macos (that have issues with MSA anyways)

**Comment by ghost** [comment]
This would also make merging absolute cancer. Not to mention that many people still have reason to use older releases.

**Comment by Scrumplex** [comment]
> This would also make merging absolute cancer.

I don't see a problem. I only needed to change two source files as they were using deprecated methods. And in the future we can just fix them on the fly while we merge upstream changes. Most of the time the changes are trivially easy

**Comment by redstrate** [comment]
I get the whole "not supporting outdated platforms" point, but what point is there in **only** supporting Qt 5.12? Is there a specific 5.12 feature that we **need**? (this is already possible, without forcing everyone to update their qt) I don't see any reason why we can't just do what we're doing already, and supporting a wide range of Qt releases while keeping tabs on what's being deprecated soon, with the move to having a Qt6 option.

**Comment by ghost** [comment]
I created PolyMC with the vision of supporting as wide a userbase as possible. Forcing Qt 5.12 cuts off such a wide range of platforms, many of which do have niche use cases. Just saying "just upgrade lol" is not something we should be telling users. As I've stated, some people still have good reason to use these older operating systems.

Because of this, I don't support forcing 5.12. We can still very easily support Qt 5.6; if there are functions only available in newer versions, use an if macro.

I'm not on board with the point about required CMake either. Prebuilt packages don't need CMake as a runtime dependency, so even if someone can't build due to QuaZip's CMake requirement, it can still be run.

And that's not to mention that people can always use an older commit of QuaZip.

So in conclusion I don't want this. The final decision is up to Lenny but it appears he is against this as well. The detriments significantly outweigh the benefits in my opinion.

**Comment by Scrumplex** [comment]
> I get the whole "not supporting outdated platforms" point, but what point is there in **only** supporting Qt 5.12? Is there a specific 5.12 feature that we **need**?

We will not be **only** supporting Qt 5.12. It is just the minimum requirement. Ideally everyone is running the latest Qt 5.15 patchset from KDE.

The main reason to bump to Qt 5.12 is to avoid bugs. By supporting a very wide range of Qt versions, we would need to workaround bugs thoughout the codebase with little gain. Qt 5.12 should be a fine baseline. We can probably remove the `iconfix` library and clean up the codebase by removing older Qt workarounds.


> Forcing Qt 5.12 cuts off such a wide range of platforms, many of which do have niche use cases.

Yes upcoming releases wouldn't run on them anymore. But users will always be able to run an older version of PolyMC (or maybe a legacy branch of the codebase?). I mean, again, why should we be supporting EOL platforms? PolyMC 1.0.6 is probably fine for this niche. Otherwise I am sure we won't merge this before 1.1.0 so they even get that.


> We can still very easily support Qt 5.6; if there are functions only available in newer versions, use an if macro.

But that makes the codevase way more complex than it should be.


> And that's not to mention that people can always use an older commit of QuaZip.

Nope. We can only use the very latest commit of upstream QuaZip. If a user wants an older version, they need to revert my PRs about QuaZip and switch to the old fork. Might as well use an older PolyMC version then.

Apart from QuaZip, we also need Java 17 for modern versions of Minecraft and MSA doesn't work on older Qt versions and it will become the ONLY option in March this year.

**Comment by d-513** [comment]
Might wanna look into launcher/DesktopServices.cpp which seems to be using werid api's like IndirectOpen which crashes when running in flatpak for example (currently flatpak is patched to not use indirectopen). It seems to be workarounds for bugs, which were most likely fixed in qt1.12

**Review by txtsd — APPROVED** [review]
LGTM!

Linux users will be getting an AppImage to run from the next release onward, unless they want to use distro based packages, so they should be fine.
Windows version comes with packaged DLLs.
macOS 10.14 is already EOL, however, Qt 5.15 supports macOS 10.13 and up, so this really isn't a problem.

EDIT: If we're really worried about users on older platforms, we can have anonymous analytics that show us what OS users are running on and make a decision that way.

**Comment by Scrumplex** [comment]
Rebased on `develop`. Added a new commit that removes `QT_VERSION_CHECK`s that aren't needed for Qt 5.12 anymore.

**Review by d-513 — APPROVED** [review]

**Comment by theofficialgman** [comment]
> debian 10 won't work for a qt issue anyways (see https://bugreports.qt.io/browse/QTBUG-68393)
> 
> so we would only miss ubuntu 18.04 (why would you use it anyways), windows xp and vista (just why) and older macos (that have issues with MSA anyways)

@DioEgizio fyi, this bug is fixed in debian buster UNRELEAED now (qt 5.11.3+dfsg1-1+deb10u5). I was waiting for a response from the maintainer and it has finally been patched (https://salsa.debian.org/qt-kde-team/qt/qtbase/-/commit/055948c03b32cdd32b54d9410f0a3dcd3d0fe651). Give it a few more days and it will release on the debian repos

**Comment by DioEgizio** [comment]
> > debian 10 won't work for a qt issue anyways (see https://bugreports.qt.io/browse/QTBUG-68393)
> > so we would only miss ubuntu 18.04 (why would you use it anyways), windows xp and vista (just why) and older macos (that have issues with MSA anyways)
> 
> @DioEgizio fyi, this bug is fixed in debian buster UNRELEAED now (qt 5.11.3+dfsg1-1+deb10u5). I was waiting for a response from the maintainer and it has finally been patched (https://salsa.debian.org/qt-kde-team/qt/qtbase/-/commit/055948c03b32cdd32b54d9410f0a3dcd3d0fe651). Give it a few more days and it will release on the debian repos

Wait over 3 years for an important patch like that

**Comment by theofficialgman** [comment]
> Wait over 3 years for an important patch like that

there was no waiting, took like a week after I notified them. The maintainer wasn't aware it affected QT 5.11 because it wasn't marked correctly by QT5 upstream. It was marked as only affecting QT 5.12 which was incorrect.

edit: apparently is was marked correctly... maybe they just missed it or though it wasn't important? not sure

**Comment by DioEgizio** [comment]
> > Wait over 3 years for an important patch like that
> 
> there was no waiting, took like a week after I notified them. The maintainer wasn't aware it affected QT 5.11 because it wasn't marked correctly by QT5 upstream. It was marked as only affecting QT 5.12 which was incorrect.
> 
> edit: apparently is was marked correctly... maybe they just missed it or though it wasn't important? not sure

oh

**Comment by Scrumplex** [comment]
(rebase)

**Comment by Scrumplex** [comment]
Can we look into this again? Ubuntu 22.04 is close, Debian 11 is fairly established now. In any case a packager could patch PolyMC to build on older Qt versions again

**Comment by d-513** [comment]
This would be nice, newer Qt is used for our builds anyway

**Comment by DioEgizio** [comment]
Yes I mean:
- for Linux users: update or use flatpak/appimage
- for Windows users: just update and our builds use newer Qt anyways
- for MacOS users: old MacOS doesn't work with MSA anyways, we use newer Qt for builds anyways

File: launcher/main.cpp
Comment (txtsd) [inline launcher/main.cpp:27]: This has an unnecessary indent now

Comment (Scrumplex) [reply launcher/main.cpp:27]: nope

Comment (txtsd) [reply launcher/main.cpp:27]: Oh woops. Those #ifdefs always throw me off.

**Comment by Scrumplex** [comment]
imagine having conflicts, cringeeee

**Comment by txtsd** [comment]
> imagine having conflicts, cringeeee

It's just the minimum cmake version.

**Comment by d-513** [comment]
can we finally get this merged for next major ver?

**Comment by DioEgizio** [comment]
I honestly think there's no reason to not merge this pr

**Review by d-513 — APPROVED** [review]

**Review by d-513 — APPROVED** [review]

**Review by d-513 — APPROVED** [review]

**Review by timoreo22 — APPROVED** [review]

File: CMakeLists.txt
Comment (flowln) [inline CMakeLists.txt]: Seems to me like a good idea to add back `-Wno-deprecated-declarations`, as without it there's a lot of warnings when building (at least on my end), especially with https://doc.qt.io/qt-5/qdir-obsolete.html#operator-eq-1

Comment (Scrumplex) [reply CMakeLists.txt]: I mean yes and no. It technically should throw warnings if something is deprecated. That's basically the point of deprecation

Comment (DioEgizio) [reply CMakeLists.txt]: ~~meanwhile CurseForge: deprecation = removal~~

Comment (flowln) [reply CMakeLists.txt]: Yes, but either we remove the things that are giving the warnings right now, or we do it some other time, and use this flag to not spam the console when building. It's 100+ warnings...

Comment (Scrumplex) [reply CMakeLists.txt]: We mostly can't remove these errors if we want to maintain compatibility with Qt5.12. One option could be to fix these warnings but use QT_VERSION_CHECK

Comment (flowln) [reply CMakeLists.txt]: It's possible... Actually, looking through it, seems like most of the errors come from the FileSystem header, so maybe there's not that many places that would need this change

Comment (Scrumplex) [reply CMakeLists.txt]: I think I have all deprecation warnings covered in #575. So I might disable deprecation stuff here, but remove the flag again in Qt 6 PR

Comment (Scrumplex) [reply CMakeLists.txt:37]: I think I have all deprecation warnings covered in #575. So I might disable deprecation stuff here, but remove the flag again in Qt 6 PR

Comment (flowln) [reply CMakeLists.txt]: Fine by me!

**Comment by Scrumplex** [comment]
(rebase)

**Review by flowln — APPROVED** [review]

**Comment by Scrumplex** [comment]
So... What are we gonna do? Just merge and forget?

**Comment by DioEgizio** [comment]
From my POV, yes, there's no reason to not do this

**Comment by DioEgizio** [comment]
Also it's necessary for #575

**Comment by txtsd** [comment]
> So... What are we gonna do? Just merge and forget?

Go for it!

**Comment by DioEgizio** [comment]
Click the "merge" button :ioa:

**Comment by Scrumplex** [comment]
YOLO

**Pull request merged**
